pimd/pim_mroute.c

```diff
@@ -504,6 +610,137 @@ int pim_mroute_msg_wrvifwhole(int fd, struct interface *ifp, const char *buf,
 	return 0;
 }
 
+#if PIM_IPV == 4
+static int process_igmp_packet(struct pim_instance *pim, const char *buf,
+			       size_t buf_size, ifindex_t ifindex)
+{
+	struct interface *ifp;
+	struct pim_interface *pim_ifp;
+	struct in_addr ifaddr;
+	struct gm_sock *igmp;
+	const struct prefix *connected_src;
+	const struct ip *ip_hdr = (const struct ip *)buf;
+
+	/* We have the IP packet but we do not know which interface this
+	 * packet was
+	 * received on. Find the interface that is on the same subnet as
+	 * the source
+	 * of the IP packet.
+	 */
+	ifp = if_lookup_by_index(ifindex, pim->vrf->vrf_id);
+
+	if (!ifp || !ifp->info)
+		return 0;
+
+	connected_src = pim_if_connected_to_source(ifp, ip_hdr->ip_src);
+
+	if (!connected_src) {
+		if (PIM_DEBUG_IGMP_PACKETS) {
+			zlog_debug(
+				"Recv IGMP packet on interface: %s from a non-connected source: %pI4",
+				ifp->name, &ip_hdr->ip_src);
+		}
+		return 0;
+	}
+
+	pim_ifp = ifp->info;
+	ifaddr = connected_src->u.prefix4;
+	igmp = pim_igmp_sock_lookup_ifaddr(pim_ifp->gm_socket_list, ifaddr);
+
+	if (PIM_DEBUG_IGMP_PACKETS) {
+		zlog_debug(
+			"%s(%s): igmp kernel upcall on %s(%p) for %pI4 -> %pI4",
+			__func__, pim->vrf->name, ifp->name, igmp,
+			&ip_hdr->ip_src, &ip_hdr->ip_dst);
+	}
+	if (igmp)
+		pim_igmp_packet(igmp, (char *)buf, buf_size);
+	else if (PIM_DEBUG_IGMP_PACKETS) {
+		zlog_debug(
+			"No IGMP socket on interface: %s with connected source: %pFX",
+			ifp->name, connected_src);
+	}
+	return 0;
+}
+#endif
+
+int pim_mroute_msg(struct pim_instance *pim, const char *buf, size_t buf_size,
+		   ifindex_t ifindex)
+{
+	struct interface *ifp;
+	const ipv_hdr *ip_hdr;
+	const kernmsg *msg;
+
+	if (buf_size < (int)sizeof(ipv_hdr))
+		return 0;
+
+	ip_hdr = (const ipv_hdr *)buf;
+
+#if PIM_IPV == 4
+	if (ip_hdr->ip_p == IPPROTO_IGMP) {
+		process_igmp_packet(pim, buf, buf_size, ifindex);
```

Comment:
Please handle error cases. If this api returns error you need to return from here and not proceed further.

Reply:
This is the last statement all other code is in elseif / and else only